src/Moq/ISetup.cs

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2007, Clarius Consulting, Manas Technology Solutions, InSTEDD.
+// All rights reserved. Licensed under the BSD 3-Clause License; see License.txt.
+
+using System.Linq.Expressions;
+
+namespace Moq
+{
+	/// <summary>
+	///   A setup configured on a mock.
+	/// </summary>
+	public interface ISetup
+	{
+		/// <summary>
+		///   The setup expression.
+		/// </summary>
+		LambdaExpression Expression { get; }
+
+		/// <summary>
+		///   Gets whether this setup is conditional.
+		/// </summary>
+		bool IsConditional { get; }
+
+		/// <summary>
+		///   Gets whether this setup was matched by at least one invocation on the mock.
+		/// </summary>
+		bool IsMatched { get; }
+
+		/// <summary>
+		///   Gets whether this setup is overridden (i.e. shadowed by a more recent setup having the same expression).
+		/// </summary>
+		bool IsOverridden { get; }
+
+		/// <summary>
+		///   Gets whether this setup is verifiable.
+		/// </summary>
+		bool IsVerifiable { get; }
+
+		/// <summary>
+		///   Checks whether this setup is a part of a composite setup.
+		///   If so, the composite setup of which this one is a part is returned via the <see langword="out"/> parameter <paramref name="originalSetup"/>.
+		///   <para>
+		///     (Composite setups are those with expressions where several member accesses or calls are chained together.
+		///     They are a much more accurate reflection of the original setup expressions you will find in your own code,
+		///     whereas the setups returned by <see cref="Mock.Setups"/> are more "technical" in nature:
+		///     They only reflect the part of an original setup expression that is of immediate concern to the mock.)
+		///   </para>
+		/// </summary>
+		/// <param name="originalSetup">
+		///   If this setup is part of a composite setup,
+		///   this <see langword="out"/> parameter will be set to the latter.
+		/// </param>
+		/// <returns>
+		///   <see langword="true"/> if this setup is a part of a composite setup;
+		///   otherwise, <see langword="false"/>.
+		/// </returns>
+		/// <seealso cref="ICompositeSetup"/>
+		/// <seealso cref="ReturnsInnerMock(out Mock)"/>
+		bool IsPartOfCompositeSetup(out ICompositeSetup originalSetup);
```

Comment:
I think this one is a bit complicated because it both gives you boolean and a value.
What makes `originalSetup` `original` -- is it more like `baseSetup`?

Would be keen the have this as two properties (or two methods) unless there are large performance implications.

Reply:
Say you have a setup with an expression that involves method chaining; `parentMock.Setup(parent => parent.Child.Member)`. Moq will split this up and create, not one, but two setups:

* a setup on `parentMock` for `parent => parent.Child` which returns some `Mock<TChild>.Object`
* a setup on that child mock for `child => child.Member`.

These are the setups you will see in the `Setups` collections of these two mocks; the original expression is gone.

This method here gives you a means to get back at the full, original, composite setup expression.

I haven't pushed this yet, but I'm most likely going to turn this method into a simple `ISetup OriginalSetup { get; }` method, which may return the very same `ISetup` instance, or a `IFluentSetup` (the new name for `ICompositeSetup`&mdash;it's somewhat closer to Moq's traditional terminology).

Reply:
Thanks, I get it now! Maybe `FullSetup` would be clearer for me, but either way it's clear now.